pkg/ccl/backupccl/system_schema.go

```diff
@@ -103,6 +113,120 @@ func jobsRestoreFunc(
 ) error {
 	executor := execCfg.InternalExecutor
 
+	const (
+		statesToRevert = `('` + string(jobs.StatusRunning) + `', ` +
+			`'` + string(jobs.StatusPauseRequested) + `', ` +
+			`'` + string(jobs.StatusPaused) + `')`
+	)
+
+	err := func() (err error) {
+		// makeJobsList writes the given jobs as a SQL tuple set to the
+		// given queryBuilder.
+		makeJobsList := func(jobIDs []int64, queryBuilder *strings.Builder) {
+			fmt.Fprint(queryBuilder, "(")
+			for i, job := range jobIDs {
+				if i > 0 {
+					fmt.Fprint(queryBuilder, ", ")
+				}
+				fmt.Fprintf(queryBuilder, "'%d'", job)
+			}
+			fmt.Fprint(queryBuilder, ")")
+		}
+
+		// Note: This will keep the payload of all non-terminal jobs in memory.
+		jobPayloads := make(map[int64]*jobspb.Payload)
+		jobsToRevert := make([]int64, 0)
+		query := `SELECT id, payload FROM ` + tempTableName + ` WHERE status IN ` + statesToRevert
+		it, err := executor.QueryIteratorEx(
+			ctx, "restore-fetching-job-payloads", txn,
+			sessiondata.InternalExecutorOverride{User: security.RootUserName()},
+			query)
+		if err != nil {
+			return errors.Wrap(err, "fetching job payloads")
+		}
+		defer func() {
+			closeErr := it.Close()
+			if err == nil {
+				err = closeErr
+			}
+		}()
+		for {
+			ok, err := it.Next(ctx)
+			if !ok {
+				if err != nil {
+					return err
+				}
+				break
+			}
+
+			r := it.Cur()
+			id, payloadBytes := r[0], r[1]
+			rawJobID, ok := id.(*tree.DInt)
+			if !ok {
+				return errors.Errorf(
+					"job: failed to read job id as DInt (was %T)", id)
+			}
+			jobID := int64(*rawJobID)
+
+			payload, err := jobs.UnmarshalPayload(payloadBytes)
+			if err != nil {
+				return errors.Wrap(err, "failed to unmarshal job to restore")
+			}
+			if payload.Type() == jobspb.TypeSchemaChange {
+				details, ok := payload.UnwrapDetails().(jobspb.SchemaChangeDetails)
+				if !ok {
+					return errors.Newf("expected schema change job to have schema change details, got %t",
+						payload.UnwrapDetails())
+				}
+				if len(payload.DescriptorIDs) != 1 {
+					continue
+				}
+				descriptorID := payload.DescriptorIDs[0]
+				desc, err := catalogkv.GetDescriptorByID(ctx, txn, execCfg.Codec, descriptorID)
+				if err != nil {
+					return err
+				}
+				table, ok := desc.(catalog.TableDescriptor)
+				if !ok {
+					continue
+				}
+				allSpans := table.PrimaryIndexSpan(execCfg.Codec)
+				for i := range details.ResumeSpanList {
+					details.ResumeSpanList[i] = jobspb.ResumeSpanList{ResumeSpans: []roachpb.Span{allSpans}}
+				}
+				payload.Details = jobspb.WrapPayloadDetails(details)
+
+				jobPayloads[jobID] = payload
```

Comment:
I wonder if we should just update the job, one by one right here, rather than buffering and risking an OOM?

Reply:
I guess details _should_ be tiny -- just the one span -- so maybe overthinking this.